### PR TITLE
File block: Add spacing support

### PR DESCRIPTION
**Pull request by carolinan**
## What?

Adds margin and padding block support to the file block.

Moves the default margin from style.scss to block.json to reduce duplicate styles:
- Before this PR: When the default margin for the block is changed in theme.json, both the margin from theme.json and style.scss is output. The style in theme.json overrides the CSS from the style.scss file.
- With this PR: The margin in theme.json replaces the margin from block.json and the duplicate style is not output.

## Why?

For consistency.

## Testing Instructions
Create a new post or page.
Add two file blocks: one should be an embedded PDF file, the second any linked file.
Confirm that there is a dimension panel and that margin and padding are available but not enabled by default.
Confirm that the default bottom margin is present.
Test the margin and padding settings in the editor and front.

Note:  The File block is not listed in Site Editor > Styles > Blocks but can be styled via theme.json.

### CODE REVIEW

**Review by aaronrobertshaw — COMMENTED** [review]
Thank you for putting this PR together @carolinan 👍 

In general, it is testing pretty well for me however, I did hit a few issues.

✅ Padding & margin work for individual blocks
✅ Spacing supports values for individual blocks are reflected in both the editor and frontend
✅ Spacing supports work regardless of whether the file is an embedded PDF or not
❓ Default bottom margin is present but is overridden by constrained layout styles
✅ Padding can be configured via theme.json and Global Styles
❓ Top/bottom margin cannot be configured via theme.json or Global Styles

The main problem appears to be the layout styles overriding the block's, whether that is the default bottom margin or top and bottom margins configured via theme.json or global styles. In the case of the theme.json/global styles margins, they are specific enough in the editor to be applied but not on the frontend.

Unfortunately, I don't have the proper context around layout supports to offer much in the way of solutions to the issues noted above. @tellthemachines or @andrewserong are the layout experts. They may have encountered this before or have some ideas 🤞 

P.S. One tiny nit, perhaps we should also note and explain the moving of the default margin style to the block.json, and the box-sizing reset, in the PR description.

**Comment by carolinan** [comment]
WordPress 6.1.1 with Twenty Twenty-Three.
Addition to theme.json:
```
			"core/file": {
				"spacing": {
					"margin": {
						"top": "200px",
						"bottom": "200px"
					}
				}
			},
```

**Site editor:**
Block is placed below the footer template part, so it is not the first block, and it is not nested.
Top and bottom margin is overridden, no matter if set in theme.json, block.json, style.scss or editor.style.scss.
Setting top and bottom margin directly on the block with the editor control works since it is inline in the style attribute on the element.

```
.editor-styles-wrapper .wp-site-blocks > * + * {
    margin-block-start: 1.5rem;
}
.editor-styles-wrapper .wp-site-blocks > * {
    margin-block-start: 0;
    margin-block-end: 0;
}
.editor-styles-wrapper .wp-block-file {
    margin-top: 200px; // Not applied, value from theme.json
    margin-bottom: 200px; // Not applied, value from theme.json
}
.wp-block-file {
    align-items: center;
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    margin-bottom: 0; // Not applied, value from editor.scss.
}
```
Front:
```
.wp-site-blocks > * {
    margin-block-start: 0;
    margin-block-end: 0;
}

.wp-block-file {
    margin-top: 200px; // Not applied, value from theme.json
    margin-bottom: 200px; // Not applied, value from theme.json.
}
```

**Block editor:**
Only one block placed, not nested. The style from theme.json is applied and visible:

image



```
.editor-styles-wrapper .edit-post-visual-editor__post-title-wrapper > :where(:not(.alignleft):not(.alignright):not(.alignfull)), .editor-styles-wrapper .block-editor-block-list__layout.is-root-container > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {
    max-width: 650px;
    margin-left: auto !important;
    margin-right: auto !important;
}

.editor-styles-wrapper .wp-block-file {
    margin-top: 200px; // Value from theme.json
    margin-bottom: 200px; // Value from theme.json
}

.editor-styles-wrapper .is-layout-constrained > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {
    max-width: var(--wp--style--global--content-size);
    margin-left: auto !important;
    margin-right: auto !important;
}

.editor-styles-wrapper .is-layout-constrained > * {
    margin-block-start: 0;
    margin-block-end: 0;
}
```
Front, viewing the post with the single file block inside: The margin is not visible.
-The post content block in the template has `is-layout-constrained`:

```
body .is-layout-constrained > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {
    max-width: var(--wp--style--global--content-size);
    margin-left: auto !important;
    margin-right: auto !important;
}

body .is-layout-constrained > * {
    margin-block-start: 0;
    margin-block-end: 0;
}
.wp-block-file {
    margin-top: 200px; // Not applied, style from theme.json
    margin-bottom: 200px; // Not applied, style from theme.json.
}
```

**Comment by aaronrobertshaw** [comment]
Looks like https://github.com/WordPress/gutenberg/issues/43404 might cover the problem with margin support styles being overridden by the layout styles.

I'm wondering if we might be better served to omit the adoption of margin support until we are closer to a solution on that issue.

**Comment by carolinan** [comment]
Yes I think you are right, let's wait.

File: packages/block-library/src/file/block.json
Comment (aaronrobertshaw) [inline packages/block-library/src/file/block.json]: Unfortunately, I don't think we can make this switch yet as it would be a regression for themes without a theme.json file due to the generated styles not being loaded for them.

See: https://github.com/WordPress/gutenberg/issues/46818

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for keeping this PR updated @carolinan 👍 

I've given it another look to see if we can unblock it but it does still appear to be blocked on two fronts.

1. The layout support styles still overriding margin support styles, currently being worked on in:
    - https://github.com/WordPress/gutenberg/pull/47399
    - https://github.com/WordPress/gutenberg/pull/47858
2. Classic themes without theme.json won't get the generated styles via the `__experimentalStyles` prop in block.json causing a regression due to the removal of the default CSS style.

If padding support would be beneficial in the short term perhaps we could just omit the margin support for now and revert the `__experimentalStyle` change.

I'm also fine with leaving the PR on the back burner until someone really needs the spacing support.

File: packages/block-library/src/file/block.json
Comment (andrewserong) [reply packages/block-library/src/file/block.json]: Thanks for mentioning this, and for linking through to the WIP work to fix layout conflicts with margins. Something that's also relevant here is that in the exploration by @tellthemachines over in https://github.com/WordPress/gutenberg/pull/47858 the Columns block's default `margin-bottom` rule has its specificity reduced. The idea in that PR is that for layout rules to be able to have reduced specificity, we likely will also need all core blocks' default margin rules to be lower specificity too (e.g. wrapped in a `:where()` rule) so that low specificity layout rules can easily override it. I imagine we'd need to do that for the file block, too.

Comment (carolinan) [reply packages/block-library/src/file/block.json]: > Unfortunately, I don't think we can make this switch yet as it would be a regression for themes without a theme.json file due to the generated styles not being loaded for them.

There is a specific stylesheet loaded for them now though? The style should be added there.

Comment (aaronrobertshaw) [reply packages/block-library/src/file/block.json]: I wasn't aware there was. I was basing my comment off both https://github.com/WordPress/gutenberg/issues/46818 and https://github.com/WordPress/gutenberg/issues/45198 which are still open indicating the issue remains.

Comment (carolinan) [reply packages/block-library/src/file/block.json]: Here is where the classic.scss file was introduced.
https://github.com/WordPress/gutenberg/pull/44334

https://developer.wordpress.org/reference/functions/wp_enqueue_classic_theme_styles/
In 6.2 it uses `wp_theme_has_theme_json() ` to determine when the file is loaded.

Comment (aaronrobertshaw) [reply packages/block-library/src/file/block.json]: Thank you for trying out the classic.scss approach @carolinan

Unfortunately, I think things have changed a bit recently. https://github.com/WordPress/gutenberg/pull/47858 landed which fixed the layout margin rule specificity. Part of that meant lowering the specificity of some blocks' CSS rules. This included the File block.

The current state of this PR means we would have the margin rule in 3 places (style.scss, classic.scss, and block.json). 

I suspect we might be better off removing the addition of the new rule to `classic.scss` and the `block.json` via `__experimentalStyle`.  The latter overrides the lower margin specificity required to work alongside layout supports.

Comment (carolinan) [reply packages/block-library/src/file/block.json]: I don't think the layout margin rules are fixed, there are issues with two of the other blocks I have spacing PR's for.

Comment (aaronrobertshaw) [reply packages/block-library/src/file/block.json]: The "fixing" of margin rules was so that a block's global styles for margin could be applied. I believe the fix from https://github.com/WordPress/gutenberg/pull/47858 is still working. It sounds like the interaction between the overall layout styles and top-level block margins is what you don't think is working correctly?

I'm about to start looking more closely at the [Preformatted spacing PR](https://github.com/WordPress/gutenberg/pull/45196). 

The left/right margins not being applied for top-level blocks makes sense but I need to have a play around with the PR to see what's happening on the frontend. I'll comment further on that PR.

File: packages/block-library/src/classic.scss
Comment (aaronrobertshaw) [inline packages/block-library/src/classic.scss]: As noted in the `__experimentalStyle` thread, we'll need to remove this style from `classic.scss` to maintain the lower specificity default set in `style.scss` required by layout supports.

File: packages/block-library/src/file/block.json
Comment (aaronrobertshaw) [reply packages/block-library/src/file/block.json]: Coming back to this one with fresh eyes, I still think we just need to remove the classic.scss style and this `__experimentalStyle` entry and we should be able to move ahead with this PR.

**Review by aaronrobertshaw — COMMENTED** [review]
This PR is pretty close 👍 

Once we remove the new style defaults that override layout styles, I believe we'll be in a position where we can merge. Any further rough edges around margins within flex containers should be covered by https://github.com/WordPress/gutenberg/pull/50825.

File: packages/block-library/src/file/block.json
Comment (carolinan) [reply packages/block-library/src/file/block.json]: Removed 👍

**Review by aaronrobertshaw — APPROVED** [review]
With the removal of the default styles overriding layout styles and work continuing on https://github.com/WordPress/gutenberg/pull/50825, this is ready to land.

Thanks for your patience @carolinan 🙇

**Pull request merged**
